internal/controller/servermaintenance_controller.go

```diff
@@ -0,0 +1,484 @@
+	}
+	if err := r.patchServerClaimAnnotation(ctx, serverClaim, claimAnnotations); err != nil {
+		return ctrl.Result{}, err
+	}
+	if serverMaintenance.Spec.Policy == metalv1alpha1.ServerMaintenancePolicyOwnerApproval {
+		claimAnnotations := serverClaim.GetAnnotations()
+		if _, ok := claimAnnotations[metalv1alpha1.ServerMaintenanceApprovalKey]; !ok {
+			log.V(1).Info("Server not approved for maintenance, waiting for approval", "Server", server.Name)
+			return ctrl.Result{}, nil
+		}
+		log.V(1).Info("Server approved for maintenance", "Server", server.Name, "Maintenance", serverMaintenance.Name)
+		if modified, err := r.patchMaintenanceState(ctx, serverMaintenance, metalv1alpha1.ServerMaintenanceStateInMaintenance); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+	}
+	if serverMaintenance.Spec.Policy == metalv1alpha1.ServerMaintenancePolicyEnforced {
+		log.V(1).Info("Enforcing maintenance", "Server", server.Name, "Maintenance", serverMaintenance.Name)
+		if modified, err := r.patchMaintenanceState(ctx, serverMaintenance, metalv1alpha1.ServerMaintenanceStateInMaintenance); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceReconciler) handleInMaintenanceState(ctx context.Context, log logr.Logger, serverMaintenance *metalv1alpha1.ServerMaintenance) (ctrl.Result, error) {
+	server, err := r.getServerRef(ctx, serverMaintenance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// put server in maintenance
+	if err := r.patchServerRef(ctx, serverMaintenance, server); err != nil {
+		return ctrl.Result{}, err
+	}
+	config, err := r.createServerBootConfiguration(ctx, serverMaintenance, server)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if config == nil {
+		if err := r.setAndPatchServerPowerState(ctx, server, serverMaintenance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStatePending || config.Status.State == "" {
+		log.V(1).Info("Server boot configuration is pending", "Server", server.Name)
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStateError {
+		if modified, err := r.patchMaintenanceState(ctx, serverMaintenance, metalv1alpha1.ServerMaintenanceStateFailed); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStateReady {
+		log.V(1).Info("Server maintenance boot configuration is ready", "Server", server.Name)
+		if err := r.setAndPatchServerPowerState(ctx, server, serverMaintenance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceReconciler) createServerBootConfiguration(ctx context.Context, maintenance *metalv1alpha1.ServerMaintenance, server *metalv1alpha1.Server) (*metalv1alpha1.ServerBootConfiguration, error) {
```

Comment:
Maybe we should call this method `applyServerBootConfiguration` as we are actually doing a `PATCH` apply here.